### PR TITLE
[optim, ckpt] feat: support Qwen3.5 MoE Muon training on Ascend (#1036)

### PR DESCRIPTION
**Pull request by Lihui-Gu**
## What does this PR do?

This PR enables **Muon optimizer support for Qwen3.5 MoE models on AscendC/NPU**, including:

- A new training YAML config for AscendC Muon training.
- Compatibility fixes for `nn.Conv1d` layers introduced by GDN when using Muon.
- A fix for `MultiOptimizer` state-dict save/restore logic when expert parallelism (EP) is disabled.
- New unit tests covering the full Qwen3.5-MoE parameter surface on meta device.

Related issue: #1036

---

### 1. New AscendC Muon Training Config

- Added `configs/multimodal/qwen3_5_moe/qwen3_5_moe_vl_muon_ascendc.yaml`.
- Configures Muon optimizer and enables `muon_expert_zero_comm: true` to eliminate all-gather communication before NS computation on MoE expert parameters.

### 2. GDN Conv1d Compatibility with Muon

- Updated parameter-classification logic in `veomni/optim/muon.py` (and related splitting logic) so that `nn.Conv1d` weights introduced by GDN are correctly routed to either Muon or AdamW.
- Ensured `split_muon_adamw_params` handles `Conv1d` weights correctly.
- Added/updated `tests/optim/test_muon_unit.py`:
  - Constructs a real Qwen3.5-MoE model on meta device.
  - Covers GDN, Attention, Router, 3D expert, shared expert, Conv1d, norm, embedding, and lm_head.
  - Validates no duplicate parameters and disjoint Muon/AdamW groups.

### 3. Fix MultiOptimizer State-Dict Handling Without EP

- **Problem:** When expert parallelism (EP) was disabled, the old code used the standard single-optimizer PyTorch paths `get_optimizer_state_dict()` / `set_optimizer_state_dict()`.
- **Impact:** Because `MultiOptimizer` internally wraps multiple sub-optimizers (Muon + AdamW), the single-optimizer path incorrectly merged or split their states, leading to missing or mismatched state entries on resume.
- **Fix:** Route state-dict load/save through `MultiOptimizer`'s own merge/split logic so that each sub-optimizer's state is aggregated and restored independently.

---

### New Config

`configs/multimodal/qwen3_5_moe/qwen3_5_moe_vl_muon_ascendc.yaml`

Key fields:

```yaml
train:
  optimizer:
    type: muon
    muon_kwargs:
      ns_implementation: gram
      gram_ns_reset_iterations: [2]
      ns_steps: 5
      ns_coefficients: [3.4445, -4.7750, 2.0315]
      muon_expert_zero_comm: true   # MoE expert zero-communication switch
```

### Launch Training

```bash
# Train from scratch
bash train.sh tasks/train_vlm.py \
    configs/multimodal/qwen3_5_moe/qwen3_5_moe_vl_muon_ascendc.yaml \
    --train.seed 42 \
    --train.enable_full_determinism true \
    --train.accelerator.ulysses_size 1 \
    --train.accelerator.ep_size 4 \
    --model.model_path Qwen/Qwen3.5-35B-A3B \
    --data.train_path coco/sharegpt4v_instruct_gpt4-vision_cap100k_coco.json \
    --data.source_name sharegpt4v_captioner_sft

# Resume from step 100
bash train.sh tasks/train_vlm.py \
    configs/multimodal/qwen3_5_moe/qwen3_5_moe_vl_muon_ascendc.yaml \
    --train.seed 42 \
    --train.enable_full_determinism true \
    --train.checkpoint.load_path ./Qwen3.5-35B-A3B-vl-muon-ascend/checkpoints/global_step_100 \
    --train.accelerator.ulysses_size 1 \
    --train.accelerator.ep_size 4 \
    --model.model_path Qwen/Qwen3.5-35B-A3B \
    --data.train_path coco/sharegpt4v_instruct_gpt4-vision_cap100k_coco.json \
    --data.source_name sharegpt4v_captioner_sft
```

---

### 1. 500-step Full Training & 100-step Resume

- 500-step full training converges normally.

飞书文档 - 图片

- Resume from step 100 aligns numerically with the full-training curve at the same coordinate.
  > (Training curve screenshots: 500-step full training / resume from step 100 / numerical alignment at the same coordinate should be attached below.)

飞书文档 - 图片 (1)

飞书文档 - 图片

### 2. MoE Expert Zero-Communication Performance

(40 MoE layers × 2 expert params: `gate_up_proj` + `down_proj`) = 80

| Item | `muon_expert_zero_comm: false` | `muon_expert_zero_comm: true` |
|---|---|---|
| Gather count / step | 80 | 0 |
| Gather time | ~872.831 ms | 0 ms |
| NS count / step | 80 | 80 |
| NS time | ~1812.774 ms | ~280.144 ms|
| Communication ratio | 32.50% | 0% |
| NS ratio | 67.50% | 100% |
| End-to-end step time | 14.678s | 12.147s |

With `muon_expert_zero_comm: true`, the NS computation is ~6.47× faster because the all-gather is eliminated, yielding 4× less data movement and a higher cache hit rate.

### 3. Unit Tests

Extended `tests/optim/test_muon_unit.py`:

- Builds a real Qwen3.5-MoE model on meta device.
- Covers:
  - GDN and Attention layers
  - Router, 3D expert weights, and shared expert weights
  - `Conv1d`, normalization, embedding, and `lm_head` parameters
  - All trainable parameters
- Verifies:
  - No duplicate parameter objects
  - Muon and AdamW parameter groups are disjoint

Extended `tests/checkpoints/test_dcp_checkpointer.py`:
- Builds a real non-EP MultiOptimizer containing DistributedMuon and AdamW.
- Covers:
  - Muon momentum state
  - AdamW first- and second-moment states
  - Single-process DCP save and restore
- Verifies:
  - Both Muon and AdamW states are included
  - Optimizer state keys are preserved
  - All tensor and scalar states are restored exactly

---

## Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks
- [x] Added/updated documentation
  - **Not applicable:** no `tasks/` scripts were moved or renamed.
- [x] Added tests to CI workflow
  - Extended `tests/optim/test_muon_unit.py`, `tests/checkpoints/test_dcp_checkpointer.py`

## Summary by CodeRabbit

* **New Features**
  * Added Qwen3.5-35B-A3B multimodal training support for Ascend/NPU hardware.
  * Added an eight-device training example with Muon optimization, expert parallelism, and video/image data.
  * Increased supported video input capacity.
  * Improved configuration of Muon optimizer settings.

* **Bug Fixes**
  * Improved distributed optimizer checkpoint saving and restoration.
  * Prevented unnecessary checkpoint state for inactive optimizers.
  * Improved AdamW handling for convolutional layers.

* **Documentation**
  * Added Ascend-based Qwen3.5 MoE training instructions.

* **Tests**
  * Added single-process and distributed checkpoint round-trip coverage.

### CODE REVIEW

File: configs/multimodal/qwen3_5_moe/qwen3_5_moe_vl_muon_ascendc.yaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    video_total_pixels: 11560550 # max_seq_len * 28^2 * 0.9, dynamic per-frame budget
+    max_frames: 16
+    fps: 2.0
+    use_audio_in_video: false
+
+train:
+  accelerator:
+    ulysses_size: 4
+    fsdp_config:
+      fsdp_mode: fsdp2
```
[the remark this answers is not in the document]
Comment (Lihui-Gu) [reply configs/multimodal/qwen3_5_moe/qwen3_5_moe_vl_muon_ascendc.yaml:34]: `docs/examples/qwen3_5.md #L238-L251` updated to use EP=4 UP=1.

**Comment by FoolPlayer** [comment]
@coderabbit review

**Review by FoolPlayer — APPROVED** [review]
